### PR TITLE
[SR-12946] Don't warn cast to unrelated metatype when involves archetypes

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
When involving generic types some protocol conformances can only be known at runtime.

cc @xedin @CodaFi 

Resolves SR-12946.

### CODE REVIEW

**Comment by CodaFi** [comment]
This looks fine to me. Thanks!

@swift-ci smoke test and merge

**Review by xedin — COMMENTED** [review]
Thank you!

**Comment by xedin** [comment]
@LucianoPAlmeida Could you check whether there are any similar SRs we could dupe to this one?

File: lib/Sema/TypeCheckConstraints.cpp [hunk 1]
```diff
@@ -3706,9 +3706,11 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
   else
     fromRequiresClass = fromType->mayHaveSuperclass();
   
-  // Casts between protocol metatypes only succeed if the type is existential.
+  // Casts between protocol metatypes only succeed if the type is existential
```
Comment (tbkka) [inline lib/Sema/TypeCheckConstraints.cpp:3709]: Is the original comment here correct?  I read the code below as saying that metatype casts always *fail* if the type is existential.

**Comment by LucianoPAlmeida** [comment]
> @LucianoPAlmeida Could you check whether there are any similar SRs we could dupe to this one?

Sure @xedin, I'll look it up :)

File: lib/Sema/TypeCheckConstraints.cpp (hunk 1 above)
Comment (LucianoPAlmeida) [reply lib/Sema/TypeCheckConstraints.cpp:3709]: Humm, it really seems like that, so I think the comment is at the very least misleading... maybe It should say that `Casts between protocol metatypes only succeed if none of the types are existential` which makes sense
cc @xedin 

Thanks for pointing that out @tbkka :)

**Comment by LucianoPAlmeida** [comment]
> @LucianoPAlmeida Could you check whether there are any similar SRs we could dupe to this one?

@xedin Found a [few](https://bugs.swift.org/browse/SR-12486?jql=status%20%3D%20open%20and%20project%20%3D%20Swift%20and%20description%20~%20%27to%20unrelated%20type%27) that may be related, but I still need to test and verify which ones involves existentials and are fixed.

**Comment by xedin** [comment]
Sounds good, thank you!

**Pull request merged**